### PR TITLE
feat(headless): add insight date facet controller

### PR DESCRIPTION
**Pull request by lbergeron**
https://coveord.atlassian.net/browse/SVCC-2206

This PR builds on top of https://github.com/coveo/ui-kit/pull/2133 and adds the date facet controller for the insight use case.

### CODE REVIEW

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.test.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
+import {buildMockDateFacetRequest} from '../../../../../test/mock-date-facet-request';
```
Comment (lbergeron) [inline packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.test.ts:1]: These tests were copied from `headless-date-filter.test.ts` keeping only tests related to dispatched actions.

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    deselectAll() {
+      coreController.deselectAll();
+      dispatch(executeSearch(logFacetClearAll(getFacetId())));
+    },
+
+    sortBy(criterion: RangeFacetSortCriterion) {
+      coreController.sortBy(criterion);
+      dispatch(
+        executeSearch(logFacetUpdateSort({facetId: getFacetId(), criterion}))
```
Comment (aelfaleh) [inline packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:60]: we need to change this event with proper client, in our case `CoveoInsightClient`

Comment (aelfaleh) [reply packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:60]: This PR add the new analytics actions https://github.com/coveo/ui-kit/pull/2147

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  DateRangeOptions,
+  DateRangeRequest,
+  DateFacetProps,
+  DateFacet,
+  DateFacetState,
+};
+export {buildDateRange};
+
+/**
+ * Creates a `DateFacet` controller instance.
```
Comment (mmitiche) [inline packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:35]:
```suggestion
 * Creates an `InsightDateFacet` controller instance.
```

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `DateFacet` controller properties.
+ * @returns A `DateFacet` controller instance.
```
Comment (mmitiche) [inline packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:39]:
```suggestion
 * @returns An `InsightDateFacet` controller instance.
```

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  DateFilterOptions,
+  DateFilterInitialState,
+  DateFilterRange,
+  DateFilterProps,
+  DateFilterState,
+  DateFilter,
+};
+
+/**
+ * Creates a `DateFilter` controller instance.
```
Comment (mmitiche) [inline packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts:34]:
```suggestion
 * Creates an `InsightDateFilter` controller instance.
```

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+ * @param engine - The headless engine.
+ * @param props - The configurable `DateFilter` controller properties.
+ * @returns A `DateFilter` controller instance.
```
Comment (mmitiche) [inline packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts:37]:
```suggestion
 * @returns An `InsightDateFilter` controller instance.
```

**Comment by mmitiche** [comment]
Are we planning to export the new insight controllers using the the `insight.index.ts` file only when all the controllers are ready?

**Comment by nathanlb** [comment]
> Are we planning to export the new insight controllers using the the `insight.index.ts` file only when all the controllers are ready?

Yep

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+ */
+export function buildInsightDateFacet(
```
Comment (btaillon) [inline packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:41]: Same as https://github.com/coveo/ui-kit/pull/2156/files#r908430309 (from another PR)

File: packages/headless/src/state/insight-app-state.ts [hunk 8]
```diff
@@ -31,5 +32,6 @@ export type InsightAppState = InsightSearchParametersState &
   FacetSection &
   FacetSearchSection &
   FacetOptionsSection &
+  DateFacetSection &
```
Comment (btaillon) [inline packages/headless/src/state/insight-app-state.ts:35]: Looks like this might belong in `InsightSearchParametersState`

**Review by btaillon — APPROVED** [review]

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts (hunk 4 above)
Comment (lbergeron) [reply packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-facet.ts:39]: Same here

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts (hunk 5 above)
Comment (lbergeron) [reply packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts:34]: Same here

File: packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts (hunk 6 above)
Comment (lbergeron) [reply packages/headless/src/controllers/insight/facets/range-facet/date-facet/headless-insight-date-filter.ts:37]: Same here

**Comment by lbergeron** [comment]
> TODO: Found an issue when using a timeframe facet with relative timeframes. It apparently caused a malformed search request yielding a 500 error. To investigate.

The issue was that the date facet request and response were not mapped correctly to translate timeframes from relative to absolute format. I modified slightly the `search-mappings` types to be able to use it for insight as well, and updated the insight execute search action accordingly.

**Review by ThibodeauJF — APPROVED** [review]

**Pull request merged**